Array Index Out of Bounds Exception - Elastic 6.8.2

On executing the following curl command:

```
curl -H'content-type:application/json' 'http://localhost:9220/_msearch?error_trace=true&pretty' -d '
{"index":["companiestable"],"ignore_unavailable":true,"preference":1572975891903}
{"aggs":{"topicFilters":{"filters":{"filters":[{"span_near":{"clauses":[{"span_term":{"description":"compani"}},{"span_term":{"description":"technolog"}}],"slop":2,"in_order":false}},{"span_near":{"clauses":[{"span_term":{"description":"market"}},{"span_term":{"description":"technolog"}}],"slop":2,"in_order":false}}]},"aggs":{"sampler":{"sampler":{"shard_size":1000},"aggs":{"subtopics":{"terms":{"field":"description","size":48,"include":["market","technolog", "compani"]}}}}}}},"size":0,"_source":{"excludes":[]},"query":{"bool":{"must":[{"match_phrase":{"description":"Real Estate"}},{"wildcard":{"description":"*"}}]}}}
'
```

we get the following error:

```
{
  "responses" : [
    {
      "took" : 12,
      "timed_out" : false,
      "_shards" : {
        "total" : 5,
        "successful" : 1,
        "skipped" : 0,
        "failed" : 4,
        "failures" : [
          {
            "shard" : 1,
            "index" : "companiestable",
            "node" : "y2dEMXn3Sd6qaR4jYByMow",
            "reason" : {
              "type" : "array_index_out_of_bounds_exception",
              "reason" : "1"
            }
          }
        ]
      },
...
```
despite the query being a valid query. This is happening for elasticsearch version 6.8.2 for [this](https://drive.google.com/drive/folders/1R_IFv-c0SqMYpZDUNHY22zQ34J2cBbbj?usp=sharing) data folder.

Thanks, from the logs, the exception seems to come from a call the SamplerAggregator makes:
```
Caused by: java.lang.ArrayIndexOutOfBoundsException: 1
        at org.elasticsearch.common.util.BigArrays$ObjectArrayWrapper.get(BigArrays.java:345) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.search.aggregations.bucket.sampler.BestDocsDeferringCollector.getDocCount(BestDocsDeferringCollector.java:286) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.search.aggregations.bucket.sampler.SamplerAggregator.buildAggregation(SamplerAggregator.java:170) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.bucketAggregations(BucketsAggregator.java:141) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.search.aggregations.bucket.filter.FiltersAggregator.buildAggregation(FiltersAggregator.java:174) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.search.aggregations.AggregationPhase.execute(AggregationPhase.java:130) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:120) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.indices.IndicesService.lambda$loadIntoContext$17(IndicesService.java:1253) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.indices.IndicesService.lambda$cacheShardLevelResult$18(IndicesService.java:1309) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.indices.IndicesRequestCache$Loader.load(IndicesRequestCache.java:164) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.indices.IndicesRequestCache$Loader.load(IndicesRequestCache.java:147) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.common.cache.Cache.computeIfAbsent(Cache.java:433) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.indices.IndicesRequestCache.getOrCompute(IndicesRequestCache.java:119) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.indices.IndicesService.cacheShardLevelResult(IndicesService.java:1315) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.indices.IndicesService.loadIntoContext(IndicesService.java:1251) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:348) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:394) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.search.SearchService.access$100(SearchService.java:126) ~[elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:359) [elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:355) [elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.search.SearchService$4.doRun(SearchService.java:1107) [elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41) [elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:751) [elasticsearch-6.8.2.jar:6.8.2]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-6.8.2.jar:6.8.2]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_151]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_151]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_151]
```
Lets see if this gives us a better idea about whats wrong.

--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/BestDocsDeferringCollector.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/BestDocsDeferringCollector.java
@@ -306,6 +306,9 @@ public class BestDocsDeferringCollector extends DeferringBucketCollector impleme
     }
 
     public int getDocCount(long parentBucket) {
+        if (perBucketSamples.size() <= parentBucket) {
+            return 0;
+        }
         PerParentBucketSamples sampler = perBucketSamples.get((int) parentBucket);
         if (sampler == null) {
             // There are conditions where no docs are collected and the aggs

--- a/server/src/test/java/org/elasticsearch/search/aggregations/bucket/sampler/SamplerAggregatorTests.java
+++ b/server/src/test/java/org/elasticsearch/search/aggregations/bucket/sampler/SamplerAggregatorTests.java
@@ -28,13 +28,20 @@ import org.apache.lucene.index.IndexWriter;
 import org.apache.lucene.index.IndexWriterConfig;
 import org.apache.lucene.index.Term;
 import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.MatchAllDocsQuery;
 import org.apache.lucene.search.TermQuery;
 import org.apache.lucene.store.Directory;
 import org.elasticsearch.index.mapper.MappedFieldType;
 import org.elasticsearch.index.mapper.NumberFieldMapper;
 import org.elasticsearch.index.mapper.TextFieldMapper;
 import org.elasticsearch.index.mapper.TextFieldMapper.TextFieldType;
+import org.elasticsearch.index.query.MatchAllQueryBuilder;
+import org.elasticsearch.index.query.MatchNoneQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
 import org.elasticsearch.search.aggregations.AggregatorTestCase;
+import org.elasticsearch.search.aggregations.bucket.filter.FiltersAggregationBuilder;
+import org.elasticsearch.search.aggregations.bucket.filter.InternalFilters;
+import org.elasticsearch.search.aggregations.bucket.terms.TermsAggregationBuilder;
 import org.elasticsearch.search.aggregations.metrics.Min;
 import org.elasticsearch.search.aggregations.metrics.MinAggregationBuilder;
 import org.elasticsearch.search.aggregations.support.AggregationInspectionHelper;
@@ -117,4 +124,36 @@ public class SamplerAggregatorTests extends AggregatorTestCase {
             }
         }
     }
+
+    /**
+     * Tests that the sampler aggregation works correctly if the parent bucket does not contain any hit.
+     */
+    public void testEmptyParentBucket() throws Exception {
+        IndexWriterConfig indexWriterConfig = newIndexWriterConfig();
+        try (Directory dir = newDirectory();
+            IndexWriter writer = new IndexWriter(dir, indexWriterConfig)) {
+
+            writer.addDocument(new Document());
+
+            try (IndexReader reader = DirectoryReader.open(writer)) {
+                IndexSearcher searcher = new IndexSearcher(reader);
+
+                QueryBuilder[] filters = new QueryBuilder[]{
+                    new MatchAllQueryBuilder(),
+                    new MatchNoneQueryBuilder()
+                };
+                FiltersAggregationBuilder samplerParent = new FiltersAggregationBuilder("filters", filters);
+                TermsAggregationBuilder samplerChild = new TermsAggregationBuilder("child").field("field");
+                SamplerAggregationBuilder sampler = new SamplerAggregationBuilder("sampler")
+                    .subAggregation(samplerChild);
+                samplerParent.subAggregation(sampler);
+
+                InternalFilters response = searchAndReduce(searcher, new MatchAllDocsQuery(), samplerParent);
+                assertEquals(response.getBuckets().size(), 2);
+                assertEquals(response.getBuckets().get(0).getDocCount(), 1);
+                assertEquals(response.getBuckets().get(1).getDocCount(), 0);
+            }
+        }
+    }
+
 }

